examples/usercall-extension-bind/runner/src/main.rs

```diff
@@ -0,0 +1,355 @@
+
+        Ok(ProxyIpV4Addr {
+            src_addr,
+            dst_addr,
+            src_port,
+            dst_port,
+        })
+    }
+}
+impl From<ProxyIpV4Addr> for (SocketAddr, SocketAddr) {
+    fn from(proxy_addr: ProxyIpV4Addr) -> (SocketAddr, SocketAddr) {
+        let src_ip_addr : Ipv4Addr = proxy_addr.src_addr.into();
+        let dst_ip_addr : Ipv4Addr = proxy_addr.dst_addr.into();
+        let local_addr = SocketAddr::new(src_ip_addr.into(), proxy_addr.src_port);
+        let peer_addr = SocketAddr::new(dst_ip_addr.into(), proxy_addr.dst_port);
+        (local_addr, peer_addr)
+    }
+}
+
+#[derive(Default)]
+struct ProxyIpV6Addr {
+    src_addr: [u8; 16],
+    dst_addr: [u8; 16],
+    src_port: u16,
+    dst_port: u16,
+}
+
+impl ProxyIpV6Addr {
+    fn from_reader(rdr: &mut Read) -> IoResult<Self> {
+        let mut addr = ProxyIpV6Addr::default(); 
+
+        let _ = rdr.read_exact(&mut addr.src_addr[..])?;
+        let _ = rdr.read_exact(&mut addr.dst_addr[..])?;
+        addr.src_port = rdr.read_u16::<NetworkEndian>()?;
+        addr.dst_port = rdr.read_u16::<NetworkEndian>()?;
+
+        Ok(addr)
+    }
+}
+
+impl From<ProxyIpV6Addr> for (SocketAddr, SocketAddr) {
+    fn from(proxy_addr: ProxyIpV6Addr) -> (SocketAddr, SocketAddr) {
+        let src_ip_addr : Ipv6Addr = proxy_addr.src_addr.into();
+        let dst_ip_addr : Ipv6Addr = proxy_addr.dst_addr.into();
+        let local_addr = SocketAddr::new(src_ip_addr.into(), proxy_addr.src_port);
+        let peer_addr = SocketAddr::new(dst_ip_addr.into(), proxy_addr.dst_port);
+        (local_addr, peer_addr)
+    }
+}
+
+enum ProxyAddrReader {
+    V4(u16),
+    V6(u16),
+    UNSPEC(u16),
+}
+
+impl ProxyAddrReader {
+    fn read(&self, rdr: &mut Read) -> IoResult<(Option<SocketAddr>, Option<SocketAddr>)> {
+       match self {
+            ProxyAddrReader::V4(len) => {
+                if *len as usize != size_of::<ProxyIpV4Addr>() {
+                    Err(Error::new(ErrorKind::InvalidData, "Unexpected address length received"))?;
+                }
+                let addr = ProxyIpV4Addr::from_reader(rdr)?;
+                let (local, peer) = addr.into();
+                Ok((Some(local), Some(peer)))
+            },
+            ProxyAddrReader::V6(len) => {
+                if *len as usize != size_of::<ProxyIpV6Addr>() {
+                    Err(Error::new(ErrorKind::InvalidData, "Unexpected address length received"))?;
+                }
+                let addr = ProxyIpV6Addr::from_reader(rdr)?;
+                let (local, peer) = addr.into();
+                Ok((Some(local), Some(peer)))
+            },
+            ProxyAddrReader::UNSPEC(len) => {
+                       let mut buf = vec![0; *len as _];
+                       let _ = rdr.read_exact(&mut buf[..])?;
+                       Ok((Some(([0, 0, 0, 0], 0).into()), Some(([0, 0, 0, 0], 0).into())))
+                      },
+        }
+    }
+}
+
+#[allow(unused)]
+struct ProxyHdrV2 {
```

Comment:
Names of this struct's elements are same as:
https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt